### ISSUE TITLE
PM2 reload cause cpu 100%

### ISSUE THREAD

**Issue by hefangshi** [Unitech/pm2#1179]
When i use pm2 reload to reload my app, i got cpu 100%

```
pm2 0.12.7
node v0.12.0
```

and when i wanna to check it with top or ps, i got this

```
-bash:fork:Resource temporarily unavailable
```

here is the pm2.log

```
2015-04-16 17:18:33: Starting execution sequence in -cluster mode- for app name:yog id:0
2015-04-16 17:18:33: App name:yog id:0 online
2015-04-16 17:18:36: -reload- New worker listening
2015-04-16 17:18:36: -reload- Old worker disconnected
2015-04-16 17:18:36: Process with pid 10083 killed
2015-04-16 17:18:36: -reload- Creating a new process waiting for connections
2015-04-16 17:18:36: Starting execution sequence in -cluster mode- for app name:yog id:1
2015-04-16 17:18:36: App name:yog id:1 online
2015-04-16 17:18:37: -reload- New worker listening
2015-04-16 17:18:37: -reload- Old worker disconnected
2015-04-16 17:18:37: Process with pid 11379 killed
2015-04-16 17:18:37: -reload- Creating a new process waiting for connections
2015-04-16 17:18:37: Starting execution sequence in -cluster mode- for app name:yog id:2
2015-04-16 17:18:37: App name:yog id:2 online
2015-04-16 17:18:39: -reload- New worker listening
2015-04-16 17:18:39: -reload- Old worker disconnected
2015-04-16 17:18:39: Process with pid 12587 killed
2015-04-16 17:18:39: -reload- Creating a new process waiting for connections
2015-04-16 17:18:39: Starting execution sequence in -cluster mode- for app name:yog id:3
2015-04-16 17:18:39: App name:yog id:3 online
2015-04-16 17:18:41: -reload- New worker listening
2015-04-16 17:18:41: -reload- Old worker disconnected
2015-04-16 17:18:41: Process with pid 14088 killed
2015-04-16 17:18:41: -reload- Creating a new process waiting for connections
2015-04-16 17:18:41: Starting execution sequence in -cluster mode- for app name:yog id:4
2015-04-16 17:18:41: App name:yog id:4 online
2015-04-16 17:18:42: -reload- New worker listening
2015-04-16 17:18:42: -reload- Old worker disconnected
2015-04-16 17:18:42: Process with pid 15434 killed
2015-04-16 17:18:42: -reload- Creating a new process waiting for connections
2015-04-16 17:18:42: Starting execution sequence in -cluster mode- for app name:yog id:5
2015-04-16 17:18:42: App name:yog id:5 online
2015-04-16 17:18:44: -reload- New worker listening
2015-04-16 17:18:44: -reload- Old worker disconnected
2015-04-16 17:18:44: Process with pid 16713 killed
2015-04-16 17:18:44: -reload- Creating a new process waiting for connections
2015-04-16 17:18:44: Starting execution sequence in -cluster mode- for app name:yog id:6
2015-04-16 17:18:44: App name:yog id:6 online
2015-04-16 17:18:46: -reload- New worker listening
2015-04-16 17:18:46: -reload- Old worker disconnected
2015-04-16 17:18:46: Process with pid 17840 killed
2015-04-16 17:18:46: -reload- Creating a new process waiting for connections
2015-04-16 17:18:46: Starting execution sequence in -cluster mode- for app name:yog id:7
2015-04-16 17:18:46: App name:yog id:7 online
2015-04-16 17:18:47: -reload- New worker listening
2015-04-16 17:18:47: -reload- Old worker disconnected
2015-04-16 17:18:47: Process with pid 19176 killed
2015-04-16 17:18:47: -reload- Creating a new process waiting for connections
2015-04-16 17:18:47: Starting execution sequence in -cluster mode- for app name:yog id:8
2015-04-16 17:18:47: App name:yog id:8 online
2015-04-16 17:18:49: -reload- New worker listening
2015-04-16 17:18:49: -reload- Old worker disconnected
2015-04-16 17:18:49: ERROR CAUGHT BY DOMAIN:
Error: The process does not have permission to send the signal to any receiving process.
    at killPid (/home/fis/nhost/node_runtime/lib/node_modules/pm2/lib/TreeKill.js:50:13)
    at /home/fis/nhost/node_runtime/lib/node_modules/pm2/lib/TreeKill.js:35:7
    at Array.forEach (native)
    at killAll (/home/fis/nhost/node_runtime/lib/node_modules/pm2/lib/TreeKill.js:27:21)
    at /home/fis/nhost/node_runtime/lib/node_modules/pm2/lib/TreeKill.js:20:7
    at isFinish (/home/fis/nhost/node_runtime/lib/node_modules/pm2/lib/TreeKill.js:65:14)
    at /home/fis/nhost/node_runtime/lib/node_modules/pm2/lib/TreeKill.js:80:16
    at ChildProcess.exithandler (child_process.js:751:5)
    at ChildProcess.emit (events.js:110:17)
    at maybeClose (child_process.js:1008:16)
2015-04-16 17:18:49: ERROR CAUGHT BY DOMAIN:
Error: spawn pm2 ENOENT
    at exports._errnoException (util.js:746:11)
    at Process.ChildProcess._handle.onexit (child_process.js:1046:32)
    at child_process.js:1137:20
    at Immediate._tickDomainCallback [as _onImmediate] (node.js:381:11)
    at processImmediate [as _immediateCallback] (timers.js:358:17)
2015-04-16 17:18:49: ERROR CAUGHT BY DOMAIN:
Error: spawn pm2 ENOENT
    at exports._errnoException (util.js:746:11)
    at Process.ChildProcess._handle.onexit (child_process.js:1046:32)
    at child_process.js:1137:20
    at Immediate._tickDomainCallback [as _onImmediate] (node.js:381:11)
    at processImmediate [as _immediateCallback] (timers.js:358:17)
2015-04-16 17:18:49: ERROR CAUGHT BY DOMAIN:
Error: spawn pm2 ENOENT
    at exports._errnoException (util.js:746:11)
    at Process.ChildProcess._handle.onexit (child_process.js:1046:32)
    at child_process.js:1137:20
    at Immediate._tickDomainCallback [as _onImmediate] (node.js:381:11)
    at processImmediate [as _immediateCallback] (timers.js:358:17)
2015-04-16 17:18:49: ERROR CAUGHT BY DOMAIN:
Error: spawn pm2 ENOENT
    at exports._errnoException (util.js:746:11)
    at Process.ChildProcess._handle.onexit (child_process.js:1046:32)
    at child_process.js:1137:20
    at Immediate._tickDomainCallback [as _onImmediate] (node.js:381:11)
    at processImmediate [as _immediateCallback] (timers.js:358:17)
2015-04-16 17:18:49: ERROR CAUGHT BY DOMAIN:
Error: spawn pm2 ENOENT
    at exports._errnoException (util.js:746:11)
    at Process.ChildProcess._handle.onexit (child_process.js:1046:32)
    at child_process.js:1137:20
    at Immediate._tickDomainCallback [as _onImmediate] (node.js:381:11)
    at processImmediate [as _immediateCallback] (timers.js:358:17)
```

and `ERROR CAUGHT BY DOMAIN` is repeat over and over

what's more, i found there is a log of defunct process for PM2:God

when i killed the pm2 daemon and restart it, everything is fine.

```
2015-04-16 17:24:18: [PM2][WORKER] Started with refreshing interval: 30000
2015-04-16 17:24:18: [[[[ PM2/God daemon launched ]]]]
2015-04-16 17:24:18: BUS system [READY] on port /home/fis/nhost/pm2/pub.sock
2015-04-16 17:24:18: RPC interface [READY] on port /home/fis/nhost/pm2/rpc.sock
2015-04-16 17:24:18: PM2 is being killed via kill method
2015-04-16 17:24:18: RPC socket closed
2015-04-16 17:24:18: PUB socket closed
2015-04-16 17:24:19: [PM2][WORKER] Started with refreshing interval: 30000
2015-04-16 17:24:19: [[[[ PM2/God daemon launched ]]]]
2015-04-16 17:24:19: BUS system [READY] on port /home/fis/nhost/pm2/pub.sock
2015-04-16 17:24:19: RPC interface [READY] on port /home/fis/nhost/pm2/rpc.sock
2015-04-16 17:24:19: Starting execution sequence in -cluster mode- for app name:yog id:0
2015-04-16 17:24:19: App name:yog id:0 online
2015-04-16 17:24:19: Starting execution sequence in -cluster mode- for app name:yog id:1
2015-04-16 17:24:19: App name:yog id:1 online
2015-04-16 17:24:19: Starting execution sequence in -cluster mode- for app name:yog id:2
2015-04-16 17:24:19: App name:yog id:2 online
2015-04-16 17:24:19: Starting execution sequence in -cluster mode- for app name:yog id:3
2015-04-16 17:24:19: App name:yog id:3 online
2015-04-16 17:24:19: Starting execution sequence in -cluster mode- for app name:yog id:4
2015-04-16 17:24:19: App name:yog id:4 online
2015-04-16 17:24:19: Starting execution sequence in -cluster mode- for app name:yog id:5
2015-04-16 17:24:19: App name:yog id:5 online
2015-04-16 17:24:19: Starting execution sequence in -cluster mode- for app name:yog id:6
2015-04-16 17:24:19: App name:yog id:6 online
2015-04-16 17:24:19: Starting execution sequence in -cluster mode- for app name:yog id:7
2015-04-16 17:24:19: App name:yog id:7 online
2015-04-16 17:24:19: Starting execution sequence in -cluster mode- for app name:yog id:8
2015-04-16 17:24:19: App name:yog id:8 online
2015-04-16 17:24:19: Starting execution sequence in -cluster mode- for app name:yog id:9
2015-04-16 17:24:19: App name:yog id:9 online
2015-04-16 17:24:19: Starting execution sequence in -cluster mode- for app name:yog id:10
2015-04-16 17:24:19: App name:yog id:10 online
2015-04-16 17:24:19: Starting execution sequence in -cluster mode- for app name:yog id:11
2015-04-16 17:24:19: App name:yog id:11 online
```

**Comment by jshkurti** [comment]
Hello, try again with PM2 v.0.12.10 please, reload has been refactored since.

**Comment by hefangshi** [comment]
OK, i'll try, but it will take time to show how this working out.

Did this problem have been reported by others?

**Comment by jshkurti** [comment]
```Error: The process does not have permission to send the signal to any receiving process.```
This one, never.

```Error: spawn pm2 ENOENT``` this one means your system is running out of memory and PM2 can't spawn new processes because of it.

**Comment by hefangshi** [comment]
As far as I know, ENOENT was used to report that `No such directory entry`, but not out of memory. 

I think the problem is pm2 didn't stop when get error on reload, it will repeat trying (with a very high frequency) which will cause the whole system is out of memory.finally. After I killed pm2 daemon, everything start getting fine. So I think at least, pm2 should get a better failover 

Also, the EPERM problem is a stange one. I'm sure that all processes in started by pm2, when this problem came out, I already use pm2 to reload my application for serveral times, so it wierd that pm2 get a EPERM when kill worker.

**Comment by jshkurti** [comment]
My bad, I read ENOMEN, you are right

**Comment by jshkurti** [comment]
I will inspect this.
In the meantime, please use pm2 0.12.10 and let me know if the bug occurs again :)

**Comment by hefangshi** [comment]
OK, thanks ;)

**Comment by hefangshi** [comment]
I assume that pm2 will read pids from $PM2_HOME/pids to kill workers, is that possible when I reload the application, pm2 write the wrong pid (or didn't update pid) for some worker, and  when I reload again, we gpt this problem? I don't think it's possible to get EPERM if the pid is correct.

**Comment by arnold-almeida** [comment]
I get this issue when i use a custom "pid_file" which is one of valid the options in the advance config.

Perhaps given that the pid is in a custom location pm2 will never find the pid on restart if its looking in $PM2_HOME/pids

Not using a custom pid location resolves this issue in my instance.

**Comment by kapv89** [comment]
Hi, any updates on this? Been facing this issue randomly, like once every 1-2 months.

Node.js Version - 7.10.0
PM2 version - 2.4.6

**Comment by Unitech** [comment]
@kapv89 please show us the logs

**Comment by kapv89** [comment]
Unfortunately, I have deleted the logs. If and when this issue arises again, I will share the logs with you.

What I have observed as to what happens when this issue arises, is that some old, un-managed processes remain alive, and they prevent newly spawned processes from binding on a particular port. This causes the new processes to crash, then pm2 tries to start them again, and they crash again, and so on.
